### PR TITLE
DAOS-1893: Create test utility to generate the new control plane config file.

### PR DESCRIPTION
**Pull request by ravalsam**
DAOS-1893: Create test utility to generate the new control plane config file.

- Server start API has been changed. Server will be started with -o .yaml
  Test will not pass any extra arguments to the server command line except must require
  which is -a (--attach_info) required further Python API to work.
- Flow:
    1> New YAML file created from base YAML file located from src/tests/ftest/data/daos_server_baseline.yml.
    2> New value can be passed from avocado_test.yaml under key "server_config" and create new test file
	   src/tests/ftest/data/daos_avocado_test.yml
	   Example, Please refer the Example file container/SimpleCreateDeleteTest.yaml
	    server_config:
		   name: daos_server
		   cpus: [0-2]
		   fabric_iface: eth0
		   env_vars:
			- ABT_ENV_MAX_NUM_XSTREAMS=100
			- ABT_MAX_NUM_XSTREAMS=100
    3> runServer(env_dict) kept as it is to work with some test like CartSelfTest.py.
- Changed "server_group" to "name" in all test yaml file to map with DAOS Control Plane name.
- Removed "/t" from daos_server_sockets.yml, Python yaml module won't work if tab is used instead of space.
- Need this Env to setup on Shell to run client side code.
   export CRT_PHY_ADDR_STR="ofi+sockets" or "ofi+psm2"
   export OFI_INTERFACE="eth0" or "ib0"
- Other variable can be set on Shell prior starting tests and those Env will get the
  higher priority by Control Plane Management.
- Verified CartSelfTest.py,SimpleCreateDeleteTest and ObjectIntegrity for sanity.

### CODE REVIEW

**Comment by ravalsam** [comment]
There is a Control Plane defect found (https://jira.hpdd.intel.com/browse/DAOS-2178) which cause 12 test cleanup failure from Functional test(DaosCoreTest)

**Comment by brianjmurrell** [comment]
> There is a Control Plane defect found (https://jira.hpdd.intel.com/browse/DAOS-2178) which cause 12 test cleanup failure from Functional test(DaosCoreTest)

So you will have to wait for that to land before you can rerun the Functional testing on this patch to get a pass?

**Comment by ravalsam** [comment]
> > There is a Control Plane defect found (https://jira.hpdd.intel.com/browse/DAOS-2178) which cause 12 test cleanup failure from Functional test(DaosCoreTest)
> 
> So you will have to wait for that to land before you can rerun the Functional testing on this patch to get a pass?

That is one option, Other option is we can change the D_LOG_FILE to default /tmp/server.log.

File: src/tests/ftest/util/ServerUtils.py [hunk 1]
```diff
@@ -33,18 +33,126 @@
...
 # a callback function used when there is cmd line I/O, not intended
 # to be used outside of this file
 def printFunc(thestring):
-        print("<SERVER>" + thestring)
+    print("<SERVER>" + thestring)
+
+def nvme_yaml_config(default_yaml_value, bdev, enabled=False):
+    """
+    Enable/Disable NVMe Mode.
+    By default it's Enabled in yaml file so disable to run with ofi+sockets.
```
Comment (skirvan44) [inline src/tests/ftest/util/ServerUtils.py]: NVMe mode and ofi+sockets, might just not know something but these don't seem related to me?  Document parameters, I don't understand the function because I'm not sure what they are.

File: src/tests/ftest/util/ServerUtils.py [hunk 2]
```diff
@@ -33,18 +33,126 @@
 import signal
 import fcntl
 import errno
+import yaml
 
 from avocado.utils import genio
 
 sessions = {}
 
+DEFAULT_YAML_FILE = "utils/config/examples/daos_server_sockets.yml"
```
Comment (skirvan44) [inline src/tests/ftest/util/ServerUtils.py]: I could see this not working at least the location when we have install packages in the future.  I guess I will see why we need these later on but I'm thinking it would be nice not to presume the test has access to these.

File: src/tests/ftest/util/ServerUtils.py [hunk 3]
```diff
@@ -33,18 +33,126 @@
...
+AVOCADO_YAML_FILE = "utils/config/daos_avocado_test.yml"
```
Comment (skirvan44) [inline src/tests/ftest/util/ServerUtils.py]: I don't see this file included here, did I miss it?  If we need such a file it should go in the tests/ftest directory.

File: src/tests/ftest/util/ServerUtils.py [hunk 4]
```diff
@@ -33,18 +33,126 @@
...
+        avocado_yaml_file: Avocado test yaml file
+        avocado_yaml_file_tmp: Avocado temporary test yaml file
+    """
+    with open(avocado_yaml_file, 'r') as rfile:
+        filedata = rfile.read()
+    filedata = filedata.replace('!mux', '')
+    with open(avocado_yaml_file_tmp, 'w') as wfile:
+        wfile.write(filedata)
+
+def update_server_options(avocado_yaml_value, default_yaml_value):
```
Comment (skirvan44) [inline src/tests/ftest/util/ServerUtils.py]: I wonder if we should just have a table built into the code that has default values we want for testing rather than reading from example file.   That would create a maintenance issue where we have to keep our default values correct, but not clear to me they will keep the example file up to date either.  Then anything in the yaml file over lays our defaults.

File: src/tests/ftest/util/ServerUtils.py [hunk 5]
```diff
@@ -33,18 +33,126 @@
...
+        with open('{}/{}'.format(basepath, DEFAULT_YAML_FILE), 'r') as read_file:
+            default_yaml_value = yaml.safe_load(read_file)
+    except Exception as excpn:
+        print("<SERVER> Exception occurred: {0}".format(str(excpn)))
+        traceback.print_exception(excpn.__class__, excpn, sys.exc_info()[2])
+        raise ServerFailed("Failed to Read {}/{}".format(basepath, DEFAULT_YAML_FILE))
+
+    #Read the values from avocado_testcase.yaml file if test ran with Avocado.
+    avocado_yaml_value = ""
+    if "AVOCADO_TEST_DATADIR" in os.environ:
```
Comment (skirvan44) [inline src/tests/ftest/util/ServerUtils.py]: Some other options would be to pass the yaml file name as a parameter or to read the test yaml in the test and load the config file relavent values as a dictionary into this function.

File: src/tests/ftest/util/ServerUtils.py (hunk 3 above)
Comment (ravalsam) [reply src/tests/ftest/util/ServerUtils.py]: This file is created during tests so it will not be available by default.

File: src/tests/ftest/util/ServerUtils.py (hunk 1 above)
Comment (ravalsam) [reply src/tests/ftest/util/ServerUtils.py]: In control plane API configuration file, utils/config/examples/*.yml. NVMe is enabled by default.
  # service with -n <path>)
  bdev_class: nvme
  bdev_list: ["0000:81:00.0"]  # generate regular nvme.conf

And we do not need this Enabled by default so this function will provide way to Enabled or Disabled that in yaml file

File: src/tests/ftest/util/ServerUtils.py (hunk 4 above)
Comment (ravalsam) [reply src/tests/ftest/util/ServerUtils.py]: May be we can copy the example file to src/test/ftest and use that as reference so we do not have to worry if example  file get changed. Thoughts?

File: src/tests/ftest/util/ServerUtils.py (hunk 2 above)
Comment (ravalsam) [reply src/tests/ftest/util/ServerUtils.py]: In one of the further comment, we may be going to refer our own copy of input.yaml  from src/ftest/ so that might resolve issue in future in case some thing get changed for install packages. Thoughts?

Comment (ravalsam) [reply src/tests/ftest/util/ServerUtils.py]: Base line yaml file created under ftest/data/daos_server_baseline.yml

File: src/tests/ftest/util/ServerUtils.py [hunk 6]
```diff
@@ -33,18 +34,110 @@
...
+                avocado_yaml_value = yaml.safe_load(read_file)
+        except Exception as excpn:
+            print("<SERVER> Exception occurred: {0}".format(str(excpn)))
+            traceback.print_exception(excpn.__class__, excpn, sys.exc_info()[2])
+            raise ServerFailed("Failed to Read {}"
+                               .format('{}.tmp'.format(avocado_yaml_file)))
+        #Remove temporary file
+        os.remove(avocado_yaml_file_tmp)
+    #Update values from avocado_testcase.yaml in DAOS yaml variables.
+    for key in avocado_yaml_value['server_config']:
```
Comment (brianjmurrell) [inline src/tests/ftest/util/ServerUtils.py]: A comment here about what you are doing with these two files to end up with a new result would be useful.

Additionally, the clarification of the names `default_value` and `avocado_yaml_value` as mentioned previously would help here.

File: src/tests/ftest/util/ServerUtils.py [hunk 7]
```diff
@@ -33,18 +34,110 @@
...
+        if key in default_value['servers'][0]:
+            default_value['servers'][0][key] = avocado_yaml_value\
+            ['server_config'][key]
+        elif key in default_value:
```
Comment (brianjmurrell) [inline src/tests/ftest/util/ServerUtils.py]: What if neither of these branches is true?  I.e. the key isn't in the global values or the `server` values in `default_value`.

Should we raise an error here about a bad file format, or is it OK for neither of these to exist?

File: src/tests/ftest/util/ServerUtils.py [hunk 8]
```diff
@@ -33,18 +34,110 @@
...
 import fcntl
 import errno
+import yaml
 
 from avocado.utils import genio
 
 sessions = {}
 
+DEFAULT_YAML_FILE = "src/tests/ftest/data/daos_server_baseline.yml"
+AVOCADO_YAML_FILE = "src/tests/ftest/data/daos_avocado_test.yml"
```
Comment (brianjmurrell) [inline src/tests/ftest/util/ServerUtils.py]: This is the DAOS server configuration file, as modified to run the Functional tests in our CI clusters, yes?

Can we rename this variable appropriately then?  AVOCADO_YAML_FILE sounds to me like the Avocado test yaml file.  IOW, the `.yaml` file that goes along with a test `.py` file.

Also, the fact that this file or the one above it are yaml files I think are immaterial to what they are and you can probably just drop YAML from variable name.

File: src/tests/ftest/daos_test/DaosCoreTest.py [hunk 9]
```diff
@@ -80,7 +81,7 @@ def tearDown(self):
                     subprocess.check_call(['ssh', host,
                                            '[ -f \"{0}\" ] && '
                                            '    mv \"{0}\" '
-                                           '    \"{1}\"'.format(logfile,
+                                           '    \"{1}\"'.format("/tmp/daos.log",
```
[the remark this answers is not in the document]
Comment (ravalsam) [reply src/tests/ftest/daos_test/DaosCoreTest.py]: Default path is /tmp/daos.log and for  CI it will copy that file in to /tmp/Functional_$TEST_TAG/.

File: src/tests/ftest/data/daos_server_baseline.yml [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+control_log_file: /tmp/daos_control.log
+
+# single server instance per config file for now
+servers:
+-
+  cpus: [0-7]               # map to -c 8, just count number of #cpus in the range for now
+  fabric_iface: eth0        # map to OFI_INTERFACE=eth0
+  fabric_iface_port: 31416  # map to OFI_PORT=31416
+  log_mask: DEBUG,RPC=ERR,MEM=ERR           # map to D_LOG_MASK=ERR
+  log_file: /tmp/daos.log # map to D_LOG_FILE=/tmp/server.log
```
[the remark this answers is not in the document]
Comment (ravalsam) [reply src/tests/ftest/data/daos_server_baseline.yml]: Default path is /tmp/daos.log and for CI it will copy that file in to /tmp/Functional_$TEST_TAG/.

File: src/tests/ftest/util/ServerUtils.py (hunk 8 above)
Comment (ravalsam) [reply src/tests/ftest/util/ServerUtils.py]: I will update the name

File: src/tests/ftest/util/ServerUtils.py [hunk 11]
```diff
@@ -33,18 +34,110 @@
...
+    This function is to create DAOS server configuration YAML file
+    based on Avocado test Yaml file.
+    Args:
+        - basepath = DAOS install basepath
+    """
+    #Read the baseline conf file data/daos_server_baseline.yml
+    try:
+        with open('{}/{}'.format(basepath,
+                                 DEFAULT_YAML_FILE), 'r') as read_file:
+            default_value = yaml.safe_load(read_file)
```
[the remark this answers is not in the document]
Comment (ravalsam) [reply src/tests/ftest/util/ServerUtils.py]: ok will change to default_value_set

File: src/tests/ftest/util/ServerUtils.py [hunk 12]
```diff
@@ -33,18 +34,110 @@
...
+    except Exception as excpn:
+        print("<SERVER> Exception occurred: {0}".format(str(excpn)))
+        traceback.print_exception(excpn.__class__, excpn, sys.exc_info()[2])
+        raise ServerFailed("Failed to Read {}/{}".format(basepath,
+                                                         DEFAULT_YAML_FILE))
+
+    #Read the values from avocado_testcase.yaml file if test ran with Avocado.
+    avocado_yaml_value = ""
+    if "AVOCADO_TEST_DATADIR" in os.environ:
```
[the remark this answers is not in the document]
Comment (ravalsam) [reply src/tests/ftest/util/ServerUtils.py]: No

File: src/tests/ftest/util/ServerUtils.py [hunk 13]
```diff
@@ -33,18 +34,110 @@
...
+        avocado_yaml_file_tmp = '{}.tmp'.format(avocado_yaml_file)
+
+        # Yaml Python module is not able to read !mux so need to
+        # remove !mux before reading.
+        remove_mux_from_yaml(avocado_yaml_file, avocado_yaml_file_tmp)
+
+        # Read avocado test yaml file.
+        try:
+            with open('{}'.format(avocado_yaml_file_tmp), 'r') as read_file:
+                avocado_yaml_value = yaml.safe_load(read_file)
```
[the remark this answers is not in the document]
Comment (ravalsam) [reply src/tests/ftest/util/ServerUtils.py]: I will change the name

File: src/tests/ftest/util/ServerUtils.py [hunk 14]
```diff
@@ -33,18 +34,110 @@
...
+    #Read the values from avocado_testcase.yaml file if test ran with Avocado.
+    avocado_yaml_value = ""
+    if "AVOCADO_TEST_DATADIR" in os.environ:
+        avocado_yaml_file = str(os.environ["AVOCADO_TEST_DATADIR"]).\
+                                split(".")[0] + ".yaml"
+        avocado_yaml_file_tmp = '{}.tmp'.format(avocado_yaml_file)
+
+        # Yaml Python module is not able to read !mux so need to
+        # remove !mux before reading.
+        remove_mux_from_yaml(avocado_yaml_file, avocado_yaml_file_tmp)
```
[the remark this answers is not in the document]
Comment (ravalsam) [reply src/tests/ftest/util/ServerUtils.py]: I have not tried but should work

File: src/tests/ftest/util/ServerUtils.py (hunk 6 above)
Comment (ravalsam) [reply src/tests/ftest/util/ServerUtils.py]: Changing the name

File: src/tests/ftest/util/ServerUtils.py (hunk 7 above)
Comment (ravalsam) [reply src/tests/ftest/util/ServerUtils.py]: It will update the new value if it's exist and if it's not it's OK to pass through.

File: src/tests/ftest/util/ServerUtils.py (hunk 1 above)
Comment (ravalsam) [reply src/tests/ftest/util/ServerUtils.py]: Ah that comment is not correct. It's not related to ofi+sockets. will remove it from the comment.

File: src/tests/ftest/util/ServerUtils.py (hunk 5 above)
Comment (ravalsam) [reply src/tests/ftest/util/ServerUtils.py]: This has been updated. If user run this test without Avocado it will use the default yaml file and start the server.

File: src/tests/ftest/util/ServerUtils.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (brianjmurrell) [reply src/tests/ftest/util/ServerUtils.py]: Sure, but in `remove_mux_from_yaml` you read the file as a string and then run `.replace('!mux', '')` on the string and then write the string into a file only to open and read it here again.  Why not just read the file here as a string, do the `.replace('!mux', '')` on the results and then read the yaml from that resulting string?

[the remark this answers is not in the document]
Comment (ravalsam) [reply src/tests/ftest/util/ServerUtils.py]: That will work too. I will update the code to use string.

File: src/tests/ftest/daos_test/DaosCoreTest.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ravalsam) [reply src/tests/ftest/daos_test/DaosCoreTest.py]: If we export D_LOG_FILE=/tmp/Functional_$TEST_TAG/daos.log, client and server log will be in the same file. How this are getting separated in CI?

[the remark this answers is not in the document]
Comment (ravalsam) [reply src/tests/ftest/daos_test/DaosCoreTest.py]: I am not sure how it worked before as I do not find client side log. With this PR client and sever logs are getting copied in same file.
For example: https://build.hpdd.intel.com/job/daos-stack/job/daos/view/change-requests/job/PR-226/35/artifact/Functional/IO%20test_daos.log.wolf-103vm9/*view*/ 

04/01-20:38:43.14 wolf-103vm9 DAOS[11118] **client** DBUG src/common/tse.c:350 register_cb() alloc(calloc) 'dtc': 40 at 0x7f976e6520b0.
04/01-20:38:43.14 wolf-103vm9 DAOS[11118] object DBUG src/object/cli_obj.c:152 obj_alloc() alloc(calloc) 'obj': 160 at 0x7f976e652fb0.

File: src/tests/ftest/util/ServerUtils.py [hunk 15]
```diff
@@ -46,6 +51,79 @@ class ServerFailed(Exception):
 def printFunc(thestring):
     print("<SERVER>" + thestring)
 
+def nvme_yaml_config(default_value_set, bdev, enabled=False):
+    """
+    Enable/Disable NVMe Mode.
+    NVMe is enabled by default in yaml file.So disable it for CI runs.
```
Comment (skirvan44) [inline src/tests/ftest/util/ServerUtils.py]: Not sure what the parameters are.

Comment (ravalsam) [reply src/tests/ftest/util/ServerUtils.py]: I will add info about the arguments and will change the function name.

File: src/tests/ftest/daos_test/DaosCoreTest.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (brianjmurrell) [reply src/tests/ftest/daos_test/DaosCoreTest.py]: > I am not sure how it worked before

The way it worked before is that the log path was set in the environment when starting the servers so that they logged to `/tmp/Functional_$TEST_TAG/daos.log` and the clients were left to log to the default path, `/tmp/daos.log`,

The key here was telling the server to log to a different file, not simply renaming the default file after the test was complete.

[the remark this answers is not in the document]
Comment (ravalsam) [reply src/tests/ftest/daos_test/DaosCoreTest.py]: This was not mention in log document so difficult to understand the behavior. So we need client log and after talking with Mohamad we can get that by passing the orterun -x D_LOG_FILE=/tmp/client.log for daos_test. So I am thinking to update the command and use that as client log

[the remark this answers is not in the document]
Comment (ravalsam) [reply src/tests/ftest/daos_test/DaosCoreTest.py]: I can see separate client log with this change.

File: src/tests/ftest/daos_test/daos_core_test.py [hunk 16]
```diff
@@ -40,6 +40,8 @@
 import write_host_file
 # pylint: enable=wrong-import-position,import-error
 
+CLIENT_LOG = "/tmp/client.log"
```
Comment (brianjmurrell) [inline src/tests/ftest/daos_test/daos_core_test.py]:
```suggestion
CLIENT_LOG = "/tmp/client_daos.log"
```

So that it gets collected and artifacted by the existing logic.

File: src/tests/ftest/daos_test/daos_core_test.py [hunk 17]
```diff
@@ -59,10 +61,12 @@ def setUp(self):
             build_paths = json.load(build_vars)
         self.basepath = os.path.normpath(build_paths['PREFIX'] + "/../")
 
-        self.server_group = self.params.get("server_group", '/server/',
+        self.server_group = self.params.get("name", '/server_config/',
                                             'daos_server')
         self.daos_test = self.basepath + '/install/bin/daos_test'
         self.orterun = self.basepath + '/install/bin/orterun'
+        #To generate the seperate client log file
+        self.orterun_env = '-x D_LOG_FILE={}'.format(CLIENT_LOG)
```
Comment (brianjmurrell) [inline src/tests/ftest/daos_test/daos_core_test.py]: This just gets the client logging to `/tmp/client.log` on the remote node.  It still needs to be collected to the local node and added to the artifacts that Jenkins collects as specified in the `Jenkinsfile`.

Since you are reworking how the logging is being done you might as well apply the same logic
https://github.com/daos-stack/daos/blob/7144ce68f1b670db343cb1b71dabc968d467f4c5/src/tests/ftest/daos_test/daos_core_test.py#L87-L96
to the client logs that is being used to collate the server logs.  This way if you get them into the same directory on the remote node as the server log is in, the existing log collection
https://github.com/daos-stack/daos/blob/7144ce68f1b670db343cb1b71dabc968d467f4c5/ftest.sh#L258-L259
and artifacting
https://github.com/daos-stack/daos/blob/7144ce68f1b670db343cb1b71dabc968d467f4c5/Jenkinsfile#L698
process will work on them.
